server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -295,7 +295,7 @@ static Metadata upgradeMetadata(Metadata metadata, IndexMetadataVerifier indexMe
         boolean changed = false;
         final Metadata.Builder upgradedMetadata = Metadata.builder(metadata);
         for (IndexMetadata indexMetadata : metadata) {
-            IndexMetadata newMetadata = indexMetadataVerifier.verifyIndexMetadata(indexMetadata, IndexVersions.MINIMUM_COMPATIBLE);
+            IndexMetadata newMetadata = indexMetadataVerifier.verifyIndexMetadata(indexMetadata, IndexVersions.MINIMUM_READONLY_COMPATIBLE);
```

Comment:
This works around issues on full restart like

```
[2024-12-18T11:03:25,298][ERROR][o.e.b.Elasticsearch      ] [test-cluster-0] fatal exception while booting Elasticsearch java.lang.IllegalStateException: The index [customers/jYuwSiThTmaPZRuj_Ewg4g] has current compatibility version [7.17.0] but the minimum compatible version is [8.0.0]. It should be re-indexed in Elasticsearch 8.x before upgrading to 9.0.0.
	at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.cluster.metadata.IndexMetadataVerifier.checkSupportedVersion(IndexMetadataVerifier.java:124)
	at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.cluster.metadata.IndexMetadataVerifier.verifyIndexMetadata(IndexMetadataVerifier.java:92)
	at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.gateway.GatewayMetaState.upgradeMetadata(GatewayMetaState.java:298)
	at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.gateway.GatewayMetaState.upgradeMetadataForNode(GatewayMetaState.java:285)
	at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.gateway.GatewayMetaState.createOnDiskPersistedState(GatewayMetaState.java:193)
	at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.gateway.GatewayMetaState.createPersistedState(GatewayMetaState.java:147)
	at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.gateway.GatewayMetaState.start(GatewayMetaState.java:105)
	at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.node.Node.start(Node.java:315)
	at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.start(Elasticsearch.java:480)
	at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.initPhase3(Elasticsearch.java:290)
	at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:81)
```

But seems to not be enough, probably also too broad of a thing to change